File: UVA/371.py
from sys import stdin

def ackermann(num):
    count = 0
    while num != 1:
        if num % 2 == 0:
            num = num // 2
        else:
            num = num * 3 + 1
        count += 1
    return count

while True:
    a, b = map(lambda x: int(x), next(stdin).strip().split(' '))
    if a == 0 or b == 0:
        break

    if a > b:
        a, b = b, a
    num = 0
    length = 0

    for i in range(a, b + 1):
        acklength = ackermann(i)
        if acklength > length:
            num = i
            length = acklength

    print('Between {} and {}, {} generates the longest sequence of {} values.'.format(
    a, b, num, length
    ))

File: UVA/371.cpp
#include<iostream>
using namespace std;

unsigned int ackermann(unsigned int num) {
  unsigned int count = 0;

  if(num == 1) {
    num = 4;
    count++;
  }

  while(num != 1) {
    if(num % 2 == 0)
      num /= 2;
    else
      num = num * 3 + 1;
    count++;
  }
  return count;
}


int main(void) {
  unsigned int a, b, num, length, i, temp;

  while(1) {
    cin >> a >> b;
    if(a == 0 && b == 0)
      goto finish;

    if(a > b) {
      temp = a;
      a = b;
      b = temp;
    }

    num = a; length = ackermann(a);

    for(i = a + 1; i <= b; ++i) {
      temp = ackermann(i);
      if(temp > length) {
        num = i;
        length = temp;
      }
    }

    cout <<
      "Between " << a  <<
      " and " << b <<
      ", " << num <<
      " generates the longest sequence of " << length <<
      " values." << endl;
  }

  finish:
  return 0;
}
